exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -493,21 +503,57 @@ public Builder withTimeService(TimeProvider timeProvider) {
     public TestKit build() {
       checkCorrectServiceNumber(services.size());
       checkCorrectValidatorNumber();
-      return newInstance(services, nodeType, validatorCount, timeProvider);
+      TestKitServiceInstances[] testKitServiceInstances = mergeServiceSpecs();
+      return newInstance(testKitServiceInstances, nodeType, validatorCount, timeServiceSpecs);
+    }
+
+    /**
+     * Turn collection of service instances into a list of
+     * {@linkplain TestKitServiceInstances} objects for native to work with.
+     */
+    private TestKitServiceInstances[] mergeServiceSpecs() {
+      checkDeployedArtifactsAreUsed();
+      return serviceArtifactFilenames.entrySet().stream()
+          .map(this::aggregateServiceSpecs)
+          .toArray(TestKitServiceInstances[]::new);
+    }
+
+    private void checkDeployedArtifactsAreUsed() {
+      Set<ServiceArtifactId> serviceIds = services.keySet();
+      Set<ServiceArtifactId> deployedArtifactIds = serviceArtifactFilenames.keySet();
+      Sets.SetView<ServiceArtifactId> unusedArtifacts =
+          Sets.difference(deployedArtifactIds, serviceIds);
+      checkArgument(unusedArtifacts.isEmpty(),
+          "Following service artifacts were deployed, but not used for service instantiation: %s",
+          unusedArtifacts.stream()
+              .map(ServiceArtifactId::toString)
+              .collect(Collectors.joining(", ")));
```

Comment:
Both SetView and ServiceArtifactIds must have proper toString with no redundant information, don't they?